cnf-certification-test/operator/util.go

```diff
@@ -0,0 +1,95 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation; either version 2 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+package operator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/test-network-function/cnf-certification-test/internal/clientsholder"
+	v1 "k8s.io/api/core/v1"
+
+	olmv1Alpha "github.com/operator-framework/api/pkg/operators/v1alpha1"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Get the list of pods mananged by the operator - retrieve from the CSV and then check the operator group to find the target namespaces to retrieve the managed pods
+func GetNonClusterOperatorCsvPods(csv *olmv1Alpha.ClusterServiceVersion) ([]*v1.Pod, error) {
+	csvName := csv.Name
+	annotations := csv.Annotations
+	operatorGroupName := annotations["olm.operatorGroup"]
+
+	// Get the operator group to get the targetNamespace
+	operatorTargetNamespaces, err := getOperatorGroupTargetNamespaceByName(operatorGroupName)
+	if err != nil {
+		return nil, fmt.Errorf("operator Group for found for %s", csvName)
+	}
+
+	var podList []*v1.Pod
+
+	for _, targetNamespace := range operatorTargetNamespaces {
+		pods, err := getAllPodsManagedByOperatorGroup(targetNamespace, operatorGroupName)
+		if err != nil {
+			continue
+		}
+		podList = append(podList, pods...)
+	}
+
+	return podList, nil
+}
+
+func getOperatorGroupTargetNamespaceByName(operatorGroupName string) ([]string, error) {
+	client := clientsholder.GetClientsHolder()
+
+	operatorGroup, err := client.OlmClient.OperatorsV1().OperatorGroups(operatorGroupName).Get(context.TODO(), operatorGroupName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	return operatorGroup.Spec.TargetNamespaces, nil
+}
+
+func getAllPodsManagedByOperatorGroup(targetNamespace, operatorGroupName string) (managedPods []*v1.Pod, err error) {
+	client := clientsholder.GetClientsHolder()
+
+	podsList, err := client.K8sClient.CoreV1().Pods(targetNamespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for index := range podsList.Items {
+		annotations := podsList.Items[index].Annotations
+		if annotations["olm.operatorGroup"] == operatorGroupName {
+			managedPods = append(managedPods, &podsList.Items[index])
+		}
+	}
+	return managedPods, nil
+}
+
+// Filter cluster operators
+func getNonClusterOperatorCsvs(allCsvs []*olmv1Alpha.ClusterServiceVersion) (nonClusterOperatorCsvs []*olmv1Alpha.ClusterServiceVersion) {
```

Comment:
Why not support operators install with all namespaces mode?

Reply:
The Operator Best Practice docs only is considering application operators. So, we assume that infrastructure operators are those which has InstallMode = AllNamespace. This is a strange assumption, but we do not know how to different infra from app operators. Hence, this logic.